### ISSUE TITLE
why My Lead OPA test add Lead fails

### ISSUE THREAD

**Issue by i042416** [i042416/KnowlegeRepository#2447]
When you perform OPA test for my lead application, test case "add Lead" via this url:
It will fail:

![clipboard1](https://user-images.githubusercontent.com/5669954/59409615-2dfd9280-8de9-11e9-82f4-457bf05b7a9d.png)

The reason could be analyzed via the below picture:

![clipboard2](https://user-images.githubusercontent.com/5669954/59409616-2dfd9280-8de9-11e9-9e8c-766be7e4bfb9.png)

Our OPA test code calls fireLiveChange function.
The live change event handling written in My lead application code is called, which performs oModel.read to read possible values for "f". Since we pass "true" to read API, this read is done asynchronously, that is, the execution flow directly returns to our OPA code.
Our OPA test code calls oInput.getSuggestionItems. Unfortunately Items is now still initial, since the asynchronous OData call does not finish yet in timeslot t1.
items[0].getText causes the failure of OPA test.
Till timeslot t2, OData request is done and success handling is called, to fill the SuggestionItems aggregation. t2 > t1, it is too late now.
For detail screenshot of the above five steps, please refer to picture below.

![clipboard3](https://user-images.githubusercontent.com/5669954/59409617-2e962900-8de9-11e9-8eaa-d5ce166eed20.png)

Screenshot for step1

![clipboard4](https://user-images.githubusercontent.com/5669954/59409619-2f2ebf80-8de9-11e9-8ea7-ca05b07a10b7.png)

Screenshot for step2

![clipboard5](https://user-images.githubusercontent.com/5669954/59409620-2f2ebf80-8de9-11e9-93cd-7c0053ef04f8.png)

Screenshot for step3

![clipboard6](https://user-images.githubusercontent.com/5669954/59409621-2f2ebf80-8de9-11e9-98ed-b486bbb731b2.png)

Screenshot for step5
# Solution
We would like to achieve the following call sequences:

![clipboard7](https://user-images.githubusercontent.com/5669954/59409623-2fc75600-8de9-11e9-9474-1c322b59d394.png)

As suggested by Diablo, we could try "wait for aggregation" concept and API to fulfill it:
http://veui5infra.dhcp.wdf.sap.corp:8080/sapui5-sdk-dist/docs/api/symbols/sap.ui.test.matchers.html
To be updated soon.

I have added the new matcher mentioned in wiki and now OPA test of add Lead test succeeds.

![clipboard8](https://user-images.githubusercontent.com/5669954/59409624-2fc75600-8de9-11e9-8aa0-b31c7a50a7c1.png)

![clipboard9](https://user-images.githubusercontent.com/5669954/59409626-2fc75600-8de9-11e9-97f2-3fc4dfd4bf82.png)

To summarize,
The newly added matcher will lead to an endless loop by a periodical check with polling interval 400 milliseconds by default, as a result the asynchronously executed OData request has enough time to finish.
In the end we have the following expected time sequence chart.

![clipboard10](https://user-images.githubusercontent.com/5669954/59409627-305fec80-8de9-11e9-864e-26e5a350aedc.png)

If you would like to debug matcher logic, please set breakpoint on line below.

![clipboard11](https://user-images.githubusercontent.com/5669954/59409628-305fec80-8de9-11e9-9b53-97dcc7a40f99.png)

要获取更多Jerry的原创文章，请关注公众号"汪子熙":
![](https://upload-images.jianshu.io/upload_images/2085791-c4f58b9813a954cb.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
